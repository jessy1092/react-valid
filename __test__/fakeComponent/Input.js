import React from 'react';

const Input = ({ validate, message }) => (
  <div>
    <input onChange={e => validate(e.target.value)} />
    <div>{message}</div>
  </div>
);

export default Input;
