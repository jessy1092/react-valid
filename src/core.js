import React from 'react';

export const addMethod = (validators, name, method = () => {}, message = '') => ({
  ...validators,
  [name]: {
    method,
    message,
  },
});

export const addMethods = (validators, appendValidators = {}) => {
  const defaultValidator = {
    method() {},
    message: '',
  };

  const newValidates = {
    ...validators,
  };

  Object.keys(appendValidators).forEach(name => {
    newValidates[name] = {
      ...defaultValidator, // Set default value to prevent method not define
      ...appendValidators[name],
    };
  });

  return newValidates;
};

export const connect = (validators, Component) => {
  const displayName = Component.displayName || Component.name || 'Component';

  class ValidComponent extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        valid: true,
        message: '',
      };

      this.validate = this.validate.bind(this);
    }

    validate(value) {
      const { onError = () => {}, ...other } = this.props;
      let message = '';

      const status = Object.keys(validators).every(name => {
        if (Object.prototype.hasOwnProperty.call(other, name)) {
          if (!validators[name].method(value, this.props[name])) {
            if (typeof validators[name].message === 'string') {
              // Display normal string message
              message = validators[name].message;
            } else {
              // Use template to generate message
              message = validators[name].message(value, this.props[name]);
            }
            return false;
          }
        }
        return true;
      });

      this.setState(
        {
          valid: status,
          message,
        },
        () => {
          if (!status) {
            onError(value);
          }
        },
      );
    }

    render() {
      return <Component {...this.props} {...this.state} validate={this.validate} />;
    }
  }
  ValidComponent.displayName = `Valle(${displayName})`;

  return ValidComponent;
};

export const template = (strings, ...keys) => (validateValue, props) => {
  const result = [strings[0]];
  keys.forEach((key, i) => {
    let value = '';
    if (key === 0) {
      value = validateValue;
    } else if (key === 1) {
      value = typeof props === 'string' || !isNaN(props) ? props : props[key];
    } else {
      value = props[key];
    }
    result.push(value, strings[i + 1]);
  });
  return result.join('');
};
