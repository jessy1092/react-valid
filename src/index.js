import { addMethod, addMethods, connect, template } from './core';

import isEmail from './validators/isEmail';
import contains from './validators/contains';
import equals from './validators/equals';
import isBoolean from './validators/isBoolean';
import isFloat from './validators/isFloat';
import isISO8601 from './validators/isISO8601';
import isInt from './validators/isInt';

export function createValle() {
  let validators = {};

  return {
    addMethod(name, method, message) {
      validators = addMethod(validators, name, method, message);
    },
    addMethods(appendValidators) {
      validators = addMethods(validators, appendValidators);
    },
    connect(Component) {
      return connect(validators, Component);
    },
    template,
  };
}

const valle = createValle();

valle.addMethods(isEmail);
valle.addMethods(contains);
valle.addMethods(equals);
valle.addMethods(isBoolean);
valle.addMethods(isFloat);
valle.addMethods(isISO8601);
valle.addMethods(isInt);

export default valle;
