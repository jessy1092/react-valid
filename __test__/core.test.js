import React from 'react';
import { mount } from 'enzyme';

import { addMethod, addMethods, connect, template, setMessages } from '../src/core';

import Input from './fakeComponent/Input';

test('It could add custom method through addMethod', () => {
  const validators = {};
  const validatorName = 'required';
  const validatorMethod = value => value !== '';
  const validatorMessage = 'It should have value';

  const newValidators = addMethod(validators, validatorName, validatorMethod, validatorMessage);

  expect(newValidators[validatorName].message).toEqual(validatorMessage);
  expect(newValidators[validatorName].method).toEqual(validatorMethod);
});

test('It could overriide default message through setMessages', () => {
  const validators = {};
  const appendValidators = {
    required: {
      method: value => value !== '',
      message: 'It should have value',
    },
  };

  const newValidators = addMethods(validators, appendValidators);
  const overrideValidators = setMessages(newValidators, {
    required: 'The message was overrided',
  });

  expect(overrideValidators.required.message).toEqual('The message was overrided');
});

test('It could add custom methods through addMethods', () => {
  const validators = {};
  const appendValidators = {
    required: {
      method: value => value !== '',
      message: 'It should have value',
    },
  };

  const newValidators = addMethods(validators, appendValidators);

  expect(newValidators).toEqual(appendValidators);
});

test('It sould be return valle component if connect the basic component', done => {
  const validators = {
    required: {
      method: value => value !== '',
      message: 'It should have value',
    },
  };

  const ValidInput = connect(validators, Input);
  const onInvalid = (wrapper, value) => {
    expect(value).toEqual('');
    expect(wrapper.state('valid')).toEqual(false);
    expect(wrapper.state('message')).toEqual('It should have value');
    expect(wrapper).toMatchSnapshot();
    done();
  };
  const wrapper = mount(<ValidInput required onInvalid={value => onInvalid(wrapper, value)} />);

  wrapper.find('input').simulate('change', {
    target: {
      value: '',
    },
  });
});

test('It sould has validatorKeys if component set validators', () => {
  const validators = {
    required: {
      method: value => value !== '',
      message: 'It should have value',
    },
  };

  const ValidInput = connect(validators, Input);
  const wrapper = mount(<ValidInput required />);

  expect(wrapper.state('validatorKeys')).toEqual(['required']);
});

test('It could pass through custom property', () => {
  const ValidInput = connect({}, Input);
  const wrapper = mount(<ValidInput className="test" />);

  expect(wrapper.find('Input').prop('className')).toEqual('test');
  expect(wrapper).toMatchSnapshot();
});

test('It could use template to generate message', () => {
  const messageGenerator = template`${0} is the default value`;

  expect(messageGenerator('default')).toEqual('default is the default value');
});

test('It could use template to generate message with property value', () => {
  const messageGenerator = template`${1} is the property value`;

  expect(messageGenerator('', 'property value')).toEqual('property value is the property value');
});

test('It could use template to generate message with property number value', () => {
  const messageGenerator = template`${1} is the property number value`;

  expect(messageGenerator('', 321)).toEqual('321 is the property number value');
});

test('It could use template to generate message with property regex value', () => {
  const messageGenerator = template`${1} is the property regex value`;

  expect(messageGenerator('', /abc/i)).toEqual('/abc/i is the property regex value');
});

test('It could use template to generate message with property object value', () => {
  const messageGenerator = template`${'foo'} is the property object value`;

  expect(messageGenerator('', { foo: 'hi' })).toEqual('hi is the property object value');
});
