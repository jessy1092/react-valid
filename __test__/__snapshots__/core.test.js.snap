// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It could pass through custom property 1`] = `
<Valle(Input)
  className="test"
>
  <Input
    className="test"
    message=""
    valid={true}
    validate={[Function]}
  >
    <div>
      <input
        onChange={[Function]}
      />
      <div />
    </div>
  </Input>
</Valle(Input)>
`;

exports[`It sould be return valle component if connect the basic component 1`] = `
<Valle(Input)
  onInvalid={[Function]}
  required={true}
>
  <Input
    message="It should have value"
    valid={false}
    validate={[Function]}
  >
    <div>
      <input
        onChange={[Function]}
      />
      <div>
        It should have value
      </div>
    </div>
  </Input>
</Valle(Input)>
`;
